### ISSUE TITLE
Installing koreader on Arch Linux

### ISSUE THREAD

**Issue by Markismus** [koreader/koreader#1552]
Building the emulator:

Installing the needed packages:
```sh
sudo pacman -S luajit libtinfo ccache tesseract-data-eng luarocks lua-sec patch wget unzip git autoconf subversion cmake nasm gcc clang sdl

sudo luarocks install busted
```
Added to ~/.bashrc:
```sh
export MACHINE='x86_64'
```

Now the command for starting the emulator doesn't work with the () and becomes:
```sh
cd koreader-$MACHINE/koreader && ./reader.lua -d ../../test
```

I now run into trouble with `json.so`:
```
/luajit: frontend/apps/reader/modules/readerdictionary.lua:7: module 'json' not found:
	no field package.preload['json']
	no file 'common/json.lua'
	no file 'rocks/share/lua/5.1/json.lua'
	no file 'frontend/json.lua'
	no file './json.lua'
	no file '/usr/local/share/luajit-2.1.0-alpha/json.lua'
	no file '/usr/local/share/lua/5.1/json.lua'
	no file '/usr/local/share/lua/5.1/json/init.lua'
	no file 'common/json.so'
	no file 'common/json.dll'
	no file '/usr/lib/lua/json.so'
	no file 'rocks/lib/lua/5.1/json.so'
	no file './json.so'
	no file '/usr/local/lib/lua/5.1/json.so'
	no file '/usr/local/lib/lua/5.1/loadall.so'
stack traceback:
	[C]: in function 'require'
	frontend/apps/reader/modules/readerdictionary.lua:7: in main chunk
	[C]: in function 'require'
	frontend/apps/reader/readerui.lua:37: in main chunk
	[C]: in function 'require'
	frontend/apps/filemanager/filemanager.lua:69: in function 'onFileSelect'
	frontend/ui/widget/filechooser.lua:185: in function 'onMenuSelect'
	frontend/ui/widget/menu.lua:293: in function 'action'
	frontend/ui/uimanager.lua:302: in function '_checkTasks'
	frontend/ui/uimanager.lua:453: in function 'handleInput'
	frontend/ui/uimanager.lua:544: in function 'run'
	./reader.lua:144: in main chunk
	[C]: at 0x00404770
```

Apparently I am running lua 5.3 and after `luarocks install json`, I now have a `json.so`  in `/usr/lib/lua/5.3/json.so`. I made a symbolic link: `sudo ln -s /usr/lib/lua/5.3/json.so /usr/lib/lua/json.so`. However now the error becomes:
```
./luajit: error loading module 'json' from file '/usr/lib/lua/json.so':
	/usr/lib/lua/json.so: undefined symbol: lua_tointegerx
stack traceback:
	[C]: at 0x00450360
	[C]: in function 'require'
	frontend/apps/reader/modules/readerdictionary.lua:7: in main chunk
	[C]: in function 'require'
	frontend/apps/reader/readerui.lua:37: in main chunk
	[C]: in function 'require'
	frontend/apps/filemanager/filemanager.lua:69: in function 'onFileSelect'
	frontend/ui/widget/filechooser.lua:185: in function 'onMenuSelect'
	frontend/ui/widget/menu.lua:293: in function 'action'
	frontend/ui/uimanager.lua:302: in function '_checkTasks'
	frontend/ui/uimanager.lua:453: in function 'handleInput'
	frontend/ui/uimanager.lua:544: in function 'run'
	./reader.lua:144: in main chunk
	[C]: at 0x00404770
```

**Comment by hwhw** [comment]
We should be self-contained, i.e. not depend on external modules. Now I'm a bit at loss where the json module is supposed to be built and provided...

**Comment by Markismus** [comment]
Couldn't find `json` in koreader nor koreader-base.

**Comment by chrox** [comment]
It's installed as a dependency when installing `lua-spore` in making koreader-base [here](https://github.com/koreader/koreader-base/blob/master/Makefile#L568-L574). So you may need to run `make clean & make` at the top of koreader src tree.

**Comment by wjchen** [comment]
Install luarocks5.1 not luarocks.

**Comment by chrox** [comment]
@wjchen That's a good point. Luarocks for lua5.3 may install `json` to `rocks/share/lua/5.3/json.lua` which is not in search path of koreader.

**Comment by Markismus** [comment]
`luarocks` installs to `/usr/lib/lua/5.3/json.so`.
`luarocks-5.1` installs to `/usr/lib/lua/5.1/json.so`.

Both are outside the path. Both install `json.so` instead of `json.lua` which is distributed with the releases.

**Comment by Markismus** [comment]
I've destroyed the git clone and started anew: with the same result. `lua-spore` is installed, but not `json.lua`, `decode.lua` and `encode.lua`.

So I've made the symlink from the file installed with luarocks-5.1:
`sudo ln -s /usr/lib/lua/5.1/json.so /usr/lib/lua/json.so`

Now it works. However, I don't think this is the way it was intended.

**Comment by NiLuJe** [comment]
Lua is a mess for PMS. FWIW, I run into roughly the same kind of issues on Gentoo.

**Comment by Markismus** [comment]
Not just lua, cross compilation is a crime, too. Instead of just installing the `gcc-arm-linux-gnueabihf` compiler for Kobo with apt-get, both the automatized yaourt (`yaourt -S gcc-arm-linux-gnueabihf`) and the [manual installation of the package](http://www.archlinuxuser.com/2013/07/how-to-install-arm-linux-gnueabi-gcc-on.html) abort on this failure:
`configure: error: Need linker with .init_array/.fini_array support.`

I am trying the linaro compiler [gcc-linaro-arm-linux-gnueabihf](https://aur.archlinux.org/packages/gcc-linaro-arm-linux-gnueabihf/) for ARM7 processors now. It gives a path error for now.

**Comment by NiLuJe** [comment]
@Markismus: That's because glibc 2.15 is terrible. You either need to use eglibc 2_15, or the 2.15 *branch* of glibc, and even then, you need to backport some more stuff from 2.16 to make it build.

cf. http://trac.ak-team.com/trac/browser/niluje/Configs/trunk/Kindle/Misc/x-compile.sh#L40

I had to jump through some hoops when ct-ng killed eglibc support, and I'm not happy with my current workaround (which is to use glibc 2.16, instead of the 2.15 branch plus relevant backports).

**Comment by Markismus** [comment]
@NiLuJe Thanks. I see you used the linaro compiler, too. It took me half an hour, but it works! I compiled the kobo update.

Install linaro armhf toolchain:
```sh
yaourt -S gcc-linaro-arm-linux-gnueabihf
```

Add the toolchain to the path. Edit bash configuration:
```sh
sudo gedit ~/.bashrc 
```
And add the line:
```sh
export PATH="/usr/lib/ccache/bin/:/opt/gcc-linaro-arm-linux-gnueabihf/bin/:$PATH"
```

For compiling I needed 32-bit tools (I got the error: `fatal error: gnu/stubs-32.h: No such file`). Enable multilib in pacman by editing the pacman configuration:
```sh
sudo gedit /etc/pacman.conf
```

and uncommenting the lines:
```
[multilib]
Include = /etc/pacman.d/mirrorlist
```
Update the installation and install the compiler and libraries:
```sh
pacman -Syu
pacman -S gcc-multilib multilib-devel
```
When asked to delete the conflicting packages, e.g. `gcc`,  choose yes. The multilib works both for x64 and 32bit.

Finally:
```sh
sudo gedit /etc/pacman.conf
```
